File: 005-lookaround-matrix/engines/probe.go
// 機能判定の Go 側。matrix.py から TSV を受け取り、判定結果を TSV で返す。
//
// Go の regexp は RE2 系で、バックトラッキングしない代わりに書けない記法がある。
// 「書けない」ことも成果物なので、コンパイルエラーはそのまま出力する。
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strings"
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	for scanner.Scan() {
		line := scanner.Text()
		if strings.TrimSpace(line) == "" {
			continue
		}
		cols := strings.Split(line, "\t")
		if len(cols) < 4 {
			continue
		}
		id, pattern, subject, expected := cols[0], cols[1], cols[2], cols[3]

		re, err := regexp.Compile(pattern)
		if err != nil {
			msg := strings.SplitN(err.Error(), "\n", 2)[0]
			fmt.Fprintf(out, "%s\tERROR:%s\n", id, msg)
			continue
		}
		got := re.FindString(subject)
		switch {
		case got == "":
			fmt.Fprintf(out, "%s\tNG:none\n", id)
		case got == expected:
			fmt.Fprintf(out, "%s\tOK\n", id)
		default:
			fmt.Fprintf(out, "%s\tNG:%s\n", id, got)
		}
	}
}

File: 005-lookaround-matrix/engines/probe.cs
// 機能判定の .NET 側。matrix.py から TSV を受け取り、判定結果を TSV で返す。
// 実行: dotnet run probe.cs（.NET 10 のファイルベース実行）
using System.Text.RegularExpressions;

string? line;
while ((line = Console.ReadLine()) != null)
{
    if (string.IsNullOrWhiteSpace(line)) continue;
    var cols = line.Split('\t');
    if (cols.Length < 4) continue;
    var (id, pattern, subject, expected) = (cols[0], cols[1], cols[2], cols[3]);

    try
    {
        var re = new Regex(pattern);
        var m = re.Match(subject);
        if (!m.Success) Console.WriteLine($"{id}\tNG:none");
        else if (m.Value == expected) Console.WriteLine($"{id}\tOK");
        else Console.WriteLine($"{id}\tNG:{m.Value}");
    }
    catch (Exception e)
    {
        var msg = e.Message.Split('\n')[0];
        Console.WriteLine($"{id}\tERROR:{msg}");
    }
}
